ext/drpttracer/drpttracer.dox

```diff
@@ -0,0 +1,130 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL VMWARE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/**
+***************************************************************************
+***************************************************************************
+\page page_drpttracer Intel PT Tracing
+
+The \p drpttracer DynamoRIO Extension provides clients with tracing functionality via
+Intel's PT instruction tracing feature.
+
+\note This extension only works on x86_64 Linux.
+
+ - \ref sec_drpttracer_setup
+ - \ref sec_drpttracer_usage
+ - \ref sec_drpttracer_tracing_mode
+
+\section sec_drpttracer_setup Setup
+
+To use \p drpttracer with your client simply include this line in your client's
+\p CMakeLists.txt file:
+
+\code use_DynamoRIO_extension(clientname drpttracer) \endcode
+
+That will automatically set up the include path and library dependence.
+
+Initialize and clean up \p drpttracer by calling drpttracer_init() and
+drpttracer_exit().
+
+\section sec_drpttracer_usage Usage
+
+Intel PT (Processor Trace) only logs control flow changes. Therefore, when decoding a
+trace, the decoder needs to get raw bits of every instruction from images.
+\p drpttracer provides APIs to use Intel PT to generate trace data and some auxiliary data
+to help the decoder (e.g. libipt) to decode the trace. One type of auxiliary data is
+sideband data. This sideband data stores perf event records, which contain the image
+change message necessary for decoding the trace. Another type of auxiliary data is the
+metadata for a trace. The metadata contains the CPU information and some other information
+that can be used to synthesize the time of PT trace and the perf event records. The
+auxiliary data enables the decoder to find the correct raw bits for every instruction.
+
+\p drpttracer provides one data structure to store all generated data:
+
+ - #drpttracer_output_t
+
+And \p drpttracer provides two APIs for client to start and stop tracing:
+
+ - drpttracer_start_tracing()
+ - drpttracer_end_tracing()
+
+To record a PT trace, the client uses drpttracer_start_tracing() to start the tracing. The
+start function lets the client specify the following parameters:
+
+ - trace_mode : The tracing mode.
+
+ - pt_size_shift: The size shift of PT trace's ring buffer. It must be greater than 0, and
+the buffer size is 2^pt_size_shift * PAGE_SIZE. The client nees to make sure the buffer
+size is large enough to store the trace.
```

Comment:
> How would a client know how large it needs to be? Is there some formula or guidance or PT docs to point at, or it should be done experimentally? 

Perf sets the PT's buffer size to 4MiB. We can set the default size as 4MiB too.
https://github.com/torvalds/linux/blob/6e7765cb477a9753670d4351d14de93f1e9dbbd4/tools/perf/Documentation/perf-intel-pt.txt#L676-L684
I add some comments to advise the user to set a buffer bigger than 4 MiB.

> it should be done experimentally?

To cover all corner cases, the user should do experiments to get the appropriate size.

> Will the client find out whether it was too small later and could re-run?

In the current version, the client can know at the end of the trace if it is too small. But the client cannot re-run the same trace. It can add the buffer size in the future.